lisp/x/profiler/callgrind.go

```diff
@@ -0,0 +1,248 @@
+package profiler
+
+import (
+	"errors"
+	"fmt"
+	"github.com/luthersystems/elps/lisp"
+	"github.com/luthersystems/elps/parser/token"
+	"os"
+	"regexp"
+	"runtime"
+	"sync"
+	"time"
+)
+
+// A profiler implementation that builds Callgrind files.
+// Heavily influenced by how XDebug works for PHP. Not because
+// I like PHP but because XDebug is very light
+// The resulting files can be opened in KCacheGrind or QCacheGrind.
+type callgrindProfiler struct {
+	sync.Mutex
+	writer      *os.File
+	runtime     *lisp.Runtime
+	enabled     bool
+	startTime   time.Time
+	refs        map[string]int
+	sourceCache map[int]*token.Location
+	refCounter  int
+	callRefs    map[int32]*callRef
+}
+
+// Returns a new Callgrind processor
+func NewCallgrindProfiler(runtime *lisp.Runtime) lisp.Profiler {
+	p := new(callgrindProfiler)
+	p.runtime = runtime
+	runtime.Profiler = p
+	return p
+}
+
+// Represents something that got called
+type callRef struct {
+	start       time.Time
+	prev        *callRef
+	name        string
+	children    []*callRef
+	duration    time.Duration
+	startMemory uint64
+	endMemory   uint64
+	file        string
+	line        int
+}
+
+func (p *callgrindProfiler) IsEnabled() bool {
+	return p.enabled
+}
+
+func (p *callgrindProfiler) Enable() error {
+	p.Lock()
+	if p.enabled {
+		return errors.New("Profiler already enabled")
+	}
+	if p.writer == nil {
+		return errors.New("No output set in profiler")
+	}
+	fmt.Fprintf(p.writer, "version: 1\ncreator: elps %s (Go %s)\n", lisp.ElpsVersion, runtime.Version())
+	fmt.Fprintf(p.writer, "cmd: %s\npart: 1\npositions: line\n\n", "TODO")
```

Comment:
just to check, is the TODO here, intentional?

Reply:
It's a reminder that I need to sort that :)